File: android/app/src/main/java/vn/petpomodoro/game/Conditions.kt
package vn.petpomodoro.game

/**
 * Conditions.kt — Registry điều kiện dùng chung (spec §1.3), combo AND/OR.
 *
 * Thêm LOẠI điều kiện mới = thêm 1 data class ở đây (đúng tinh thần
 * `registerConditionType(key, fn)` của spec — Kotlin sealed interface
 * thay cho bảng COND bằng key), KHÔNG phải sửa hệ cũ.
 *
 * Mọi điều kiện biết:
 *  - [evaluate]  chống lên [GameState] (core §1.1 + trường hệ đăng ký),
 *  - [describe]  chuỗi tiếng Việt cho UI (thẻ nhánh, checklist Stats…),
 *  - [leaves]    các lá để UI vẽ progress "62/85" như demo KB1.
 */
sealed interface Condition {

    fun evaluate(s: GameState): Boolean

    fun describe(): String

    /** Các điều kiện lá (kết quả tính hiện tại + mục tiêu) cho progress bar */
    fun leaves(s: GameState): List<LeafProgress>

    data class LevelAtLeast(val value: Int) : Condition {
        override fun evaluate(s: GameState) = s.level >= value
        override fun describe() = "Lv ≥ $value"
        override fun leaves(s: GameState) =
            listOf(LeafProgress(describe(), s.level, value))
    }

    data class PowerAtLeast(val value: Int) : Condition {
        override fun evaluate(s: GameState) = s.power >= value
        override fun describe() = "Sức mạnh ≥ $value"
        override fun leaves(s: GameState) =
            listOf(LeafProgress(describe(), s.power, value))
    }

    data class StreakAtLeast(val value: Int) : Condition {
        override fun evaluate(s: GameState) = s.streak >= value
        override fun describe() = "streak ≥ $value ngày"
        override fun leaves(s: GameState) =
            listOf(LeafProgress(describe(), s.streak, value))
    }

    data class HealthyDaysAtLeast(val value: Int) : Condition {
        override fun evaluate(s: GameState) = s.healthyDays >= value
        override fun describe() = "Khỏe mạnh liên tục ≥ $value ngày"
        override fun leaves(s: GameState) =
            listOf(LeafProgress(describe(), s.healthyDays, value))
    }

    data class SessionsTodayAtLeast(val value: Int) : Condition {
        override fun evaluate(s: GameState) = s.todaySessions >= value
        override fun describe() = "phiên hôm nay ≥ $value"
        override fun leaves(s: GameState) =
            listOf(LeafProgress(describe(), s.todaySessions, value))
    }

    data class And(val of: List<Condition>) : Condition {
        override fun evaluate(s: GameState) = of.all { it.evaluate(s) }
        override fun describe() = of.joinToString(" ∧ ") { it.describe() }
        override fun leaves(s: GameState) = of.flatMap { it.leaves(s) }
    }

    data class Or(val of: List<Condition>) : Condition {
        override fun evaluate(s: GameState) = of.any { it.evaluate(s) }
        override fun describe() = of.joinToString(" ∨ ") { it.describe() }
        override fun leaves(s: GameState) = of.flatMap { it.leaves(s) }
    }
}

/** Tiến độ 1 điều kiện lá, render thành progress bar như demo KB1 (62/85 · 7/10) */
data class LeafProgress(val label: String, val current: Int, val target: Int) {
    val fraction: Float get() = if (target <= 0) 1f else (current.toFloat() / target).coerceIn(0f, 1f)
    val met: Boolean get() = current >= target
}

File: web/shared/game/conditions.ts
/**
 * PetPomodoro — Registry điều kiện dùng chung (spec §1.3, combo AND/OR).
 * Thêm loại mới = đăng ký 1 key, KHÔNG sửa hệ cũ.
 * Ví dụ mở rộng (spec §1.6):
 *   COND.total_sessions_at_least = (s, c) => s.totalSessions >= (c.value ?? 0)
 */
import type { GameState } from './types'

export type CondType =
  | 'level_at_least'
  | 'power_at_least'
  | 'streak_at_least'
  | 'healthy_days_at_least'
  | 'sessions_today_at_least'
  | 'and'
  | 'or'

export interface Cond {
  type: CondType
  value?: number
  of?: Cond[]
}

/** View "phẳng" fed vào registry — gộp core + state các hệ. */
export interface CondView {
  level: number
  power: number
  streak: number
  healthyDays: number
  todaySessions: number
  totalSessions: number
}

export function condView(state: GameState): CondView {
  return {
    level: state.pet.level,
    power: state.systems.power.value,
    streak: state.pet.streak,
    healthyDays: state.systems.stats.healthyDays,
    todaySessions: state.pet.todaySessions,
    totalSessions: state.systems.stats.totalSessions,
  }
}

export const COND: Record<CondType, (view: CondView, c: Cond) => boolean> = {
  level_at_least: (v, c) => v.level >= (c.value ?? 0),
  power_at_least: (v, c) => v.power >= (c.value ?? 0),
  streak_at_least: (v, c) => v.streak >= (c.value ?? 0),
  healthy_days_at_least: (v, c) => v.healthyDays >= (c.value ?? 0),
  sessions_today_at_least: (v, c) => v.todaySessions >= (c.value ?? 0),
  and: (v, c) => (c.of ?? []).every((x) => evalCond(v, x)),
  or: (v, c) => (c.of ?? []).some((x) => evalCond(v, x)),
}

export function evalCond(view: CondView, c: Cond): boolean {
  return COND[c.type](view, c)
}

/** Text tiếng Việt thân thiện cho điều kiện (UI render từ registry, không hard-code). */
export function describeCond(c: Cond): string {
  switch (c.type) {
    case 'level_at_least': return `Đạt Lv.${c.value} trở lên`
    case 'power_at_least': return `Sức mạnh ${c.value ?? 0}+`
    case 'streak_at_least': return `Streak ${c.value ?? 0} ngày+`
    case 'healthy_days_at_least': return `Khỏe mạnh ${c.value} ngày liền`
    case 'sessions_today_at_least': return `${c.value} phiên hôm nay`
    case 'and': return (c.of ?? []).map(describeCond).join(' và ')
    case 'or': return (c.of ?? []).map(describeCond).join(' hoặc ')
  }
}

/** Tiến độ đạt điều kiện (cur/target) — cho progress bar. Với and/or: lấy con tiến nhất. */
export function condProgress(view: CondView, c: Cond): { cur: number; target: number } {
  switch (c.type) {
    case 'level_at_least': return { cur: view.level, target: c.value ?? 0 }
    case 'power_at_least': return { cur: view.power, target: c.value ?? 0 }
    case 'streak_at_least': return { cur: view.streak, target: c.value ?? 0 }
    case 'healthy_days_at_least': return { cur: view.healthyDays, target: c.value ?? 0 }
    case 'sessions_today_at_least': return { cur: view.todaySessions, target: c.value ?? 0 }
    case 'and':
    case 'or': {
      const kids = (c.of ?? []).map((k) => condProgress(view, k))
      if (!kids.length) return { cur: 1, target: 1 }
      const ratio = (p: { cur: number; target: number }) => p.cur / Math.max(1, p.target)
      return kids.reduce((a, b) => (ratio(b) > ratio(a) ? b : a))
    }
  }
}

/* ---------- Helper dựng điều kiện (catalog đọc cho gọn) ---------- */
export const cLevel = (v: number): Cond => ({ type: 'level_at_least', value: v })
export const cPower = (v: number): Cond => ({ type: 'power_at_least', value: v })
export const cStreak = (v: number): Cond => ({ type: 'streak_at_least', value: v })
export const cHealthyDays = (v: number): Cond => ({ type: 'healthy_days_at_least', value: v })
export const cSessionsToday = (v: number): Cond => ({ type: 'sessions_today_at_least', value: v })
export const cAnd = (...of: Cond[]): Cond => ({ type: 'and', of })
export const cOr = (...of: Cond[]): Cond => ({ type: 'or', of })
